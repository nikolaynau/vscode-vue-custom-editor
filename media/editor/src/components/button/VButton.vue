<script setup lang="ts">
import { ref, computed, toRefs } from 'vue';

export interface Props {
  tag?: string;
  tabIndex?: number;
  disabled?: boolean;
}

const props = withDefaults(defineProps<Props>(), {
  tag: 'a',
  tabIndex: 0,
  disabled: false
});

const emit = defineEmits<{
  (e: 'click', ev: MouseEvent | KeyboardEvent): void;
}>();

const { disabled } = toRefs(props);
const button = ref<HTMLElement | null>(null);

const cssClasses = computed(() => ({
  'v-button--disabled': disabled.value
}));

function onClick(e: MouseEvent | KeyboardEvent) {
  emit('click', e);
}

function onEscKey() {
  button.value?.blur();
}
</script>

<template>
  <component
    ref="button"
    :is="tag"
    class="v-button"
    :class="cssClasses"
    :tabindex="disabled ? null : tabIndex"
    @click="disabled ? null : onClick($event)"
    @keyup.esc="disabled ? null : onEscKey()"
    @keyup.space.enter="disabled ? null : onClick($event)"
  >
    <slot></slot>
  </component>
</template>

<style>
@import 'button';
</style>
