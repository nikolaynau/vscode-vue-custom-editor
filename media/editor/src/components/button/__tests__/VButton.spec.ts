import { describe, it, expect } from 'vitest';
import { mount } from '@vue/test-utils';
import VButton from '../VButton.vue';

describe('VButton', () => {
  it('default render props', () => {
    const wrapper = mount(VButton);
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('tag prop', () => {
    const wrapper = mount(VButton, { props: { tag: 'button' } });
    expect(wrapper.element.tagName).toBe('BUTTON');
  });

  it('tabIndex prop', () => {
    const wrapper = mount(VButton, { props: { tabIndex: 1 } });
    expect(wrapper.attributes('tabindex')).toBe('1');
  });

  it('disabled prop', () => {
    const wrapper = mount(VButton, { props: { disabled: true } });
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('default slot', () => {
    const wrapper = mount(VButton, { slots: { default: 'Button' } });
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('click event', () => {
    const wrapper = mount(VButton, {
      props: { tag: 'button', type: 'button' }
    });
    wrapper.findComponent('button').trigger('click');
    const clickEvent = wrapper.emitted('click');
    expect(clickEvent).toHaveLength(1);
  });

  it('no click event when button is disabled', () => {
    const wrapper = mount(VButton, {
      props: { tag: 'button', type: 'button', disabled: true }
    });
    wrapper.findComponent('button').trigger('click');
    const clickEvent = wrapper.emitted('click');
    expect(clickEvent).toBeUndefined();
  });
});
