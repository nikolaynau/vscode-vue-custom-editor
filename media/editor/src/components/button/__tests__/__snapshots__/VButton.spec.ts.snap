// Vitest Snapshot v1

exports[`VButton > default slot 1`] = `"<a class=\\"v-button\\" tabindex=\\"0\\">Button</a>"`;

exports[`VButton > disabled prop 1`] = `"<a class=\\"v-button v-button--disabled\\"></a>"`;

exports[`VButton > render default props 1`] = `"<a class=\\"v-button\\" tabindex=\\"0\\"></a>"`;
